[TF:XLA] Add cluster_scoping_pass.

trentlo opened the pull request:
@sanjoy, could you please take a look of this PR when you are free. Thanks.

This new cluster_scoping_pass uses some heuristics to add scopes to nodes to guide the clustering results. Currently, the only heuristic is to preserve the parallelism between Tensorflow pipeline stages.

Some notes:
1. This PR resolves a clustering issue in a NVIDIA GoogleNet implementation due to bad clustering adding artificial dependencies between input pipeline stages. The performance impact is significant. XLA was 20% slower than Classic TF (without the PR) and becomes 20% faster with this PR.
2. This PR changes the effects of the JIT scopes such that the scopes are respected by auto-jit (These scopes were overridden/ignored by auto-jit). It feels to me a good behavior, but please let me know whether you have any concerns about it.

trentlo committed:
[TF:XLA] Add cluster_scoping_pass.

This pass uses some heuristics to add scopes to nodes to guide the
clustering results. Currently, the only heuristic is to preserve the
parallelism between Tensorflow pipeline stages.

sanjoy commented on tensorflow/compiler/jit/mark_for_compilation_pass_test.cc:
```diff
@@ -465,11 +465,8 @@ TEST(XlaCompilationTest, CyclesWithAllDifferentScopesGlobalJitOverridden) {
-  EXPECT_EQ(3, clusters.size());
-  EXPECT_EQ(clusters["A"], clusters["B"]);
-  EXPECT_EQ(clusters["A"], clusters["C"]);
+  // In this case, the GlobalJitLevel respects the scopes to cluster.
```
Preserving this behavior is important.  We have internal users that rely on `--tf_xla_auto_jit=2` to create maximally large clusters without paying any attention to jit scopes.  Can we change `ClusterScopingPass` to use a different attribute that is always respected by `MarkForCompilationPass` and leave our handling of jit scopes the same?

trentlo replied:
Got it. No problem. Will add a new attribute. Probably name it `_XlaAutoScope`?

sanjoy commented on tensorflow/compiler/jit/cluster_scoping_pass_test.cc:
```diff
@@ -0,0 +1,172 @@
+  return pass.Run(opt_options);
+}
+
+std::unordered_map<string, string> GetXlaScopes(const Graph& graph) {
```
Use `absl::flat_hash_map`.

Usually in new code we prefer using ABSL versions whenever possible.  There is some old code that uses std versions of maps and sets because originally TF did not include ABSL.

sanjoy commented on tensorflow/compiler/jit/cluster_scoping_pass.cc:
```diff
@@ -0,0 +1,152 @@
+  //
+  // Unstage -> Node_Y
+  //
+  TF_RETURN_IF_ERROR(ScopingForPipelineStages());
```
This can be just `return ScopingForPipelineStages();`.

sanjoy commented on tensorflow/compiler/jit/cluster_scoping_pass.cc:
```diff
@@ -0,0 +1,152 @@
+  Graph* graph = options.graph->get();
+
+  TF_RETURN_IF_ERROR(
+      ClusterScopingPassImpl{graph, GetGlobalJitLevelForGraph(options)}.Run());
```
This can be just `return ClusterScopingPassImpl{graph, GetGlobalJitLevelForGraph(options)}.Run();`.

sanjoy commented on tensorflow/compiler/jit/cluster_scoping_pass.cc:
```diff
@@ -0,0 +1,152 @@
+ private:
+  Graph* graph_;
+  OptimizerOptions::GlobalJitLevel global_jit_level_;
+  size_t unique_scope_id;
```
s/unique_scope_id/unique_scope_id_/

sanjoy reviewed (changes requested):
Didn't yet do an in-depth review, but leaving some nitpicky comments after a first scan.

trentlo reviewed (commented):
Thanks for the quick review. If appropriate, let me update the code after you finish your first round of review.

trentlo committed:
Address review comments by polishing coding style.

trentlo committed:
Revert the change to still override the _XlaScope when auto_jit is on.

rthadur commented:
@trentlo can you please resolve conflicts ?

trentlo committed:
Merge branch 'master' into github-cluster-scoping

trentlo committed:
Add a new attribute _XlaAutoJitScope for cluster scoping.

In other words, do not use the _XlaScope attribute, which has been used
when auto_jit is off. In this way, we do not change any existing clustering
behaviors and the new attribute is used only when auto_jit is on.

trentlo committed:
[TF:XLA] Polish naming and comments for cluster_scoping_pass.

trentlo commented:
> @trentlo can you please resolve conflicts ?

Sure. Done.

trentlo commented:
@sanjoy: I should have addressed all of your existing comments. Please help to take a look again when you have a moment. Thanks.

sanjoy commented on tensorflow/compiler/jit/mark_for_compilation_pass.cc:
```diff
@@ -923,20 +923,35 @@ MarkForCompilationPassImpl::ClusteringWillIntroduceInterDeviceDependency(
+  // The difference between _XlaScope and _XlaAutoJitScope is that _XlaScope is
+  // provided by users through jit_scope APIs, while _XlaAutoJitScope is
+  // automatically generated by the ClusterScopingPass when auto_jit is on.  As
+  // such, we respect _kXlaScope only when auto_jit is off, while respecting
```
Either s/_kXlaScope/_XlaScope/ or s/_kXlaScope/kXlaScope/

Same for _kXlaAutoJitScope

sanjoy commented on tensorflow/compiler/jit/defs.h:
```diff
@@ -24,6 +24,7 @@ namespace tensorflow {
 // Name of attribute used to tag operators for compilation with XLA
 extern const char* const kXlaCompileAttr;  // "_XlaCompile"
 extern const char* const kXlaScopeAttr;    // "_XlaScope"
+extern const char* const kXlaAutoJitScopeAttr;  // "_XlaAutoJitScope"
```
I think this should have "internal" somewhere in its name to make it obvious that it isn't for end users.  What do you think?

trentlo replied:
Good point. Let's call it `_XlaInternalScope`. It is indeed clearer than `_XlaAutoJitScope` and more distinguishable from _XlaScope.

sanjoy commented on tensorflow/compiler/jit/cluster_scoping_pass.cc:
```diff
@@ -0,0 +1,148 @@
+  SetXlaAutoJitScope(node, updated_scope);
+}
+
+void ClusterScopingPassImpl::AddScopeToAllPredecessors(Node* start) {
```
This should be called `AddScopeToAllTransitivePredecessors`.  Same for `AddScopeToAllSuccessors`.

trentlo replied:
Agreed.

sanjoy commented on tensorflow/compiler/jit/cluster_scoping_pass.cc:
```diff
@@ -0,0 +1,148 @@
+    return Status::OK();
+  }
+
+  // This preserves the parallelism between pipeline stages.  For example,
```
This comment should live in `ScopingForPipelineStages` I think.

trentlo replied:
Agreed.

sanjoy commented on tensorflow/compiler/jit/cluster_scoping_pass_test.cc:
```diff
@@ -0,0 +1,182 @@
+  return opts.FinalizeBuilder(&node_builder);
+}
+
+TEST(XlaCompilationTest, StagePipelinePreserved) {
```
How about copy-pasting this test case also into the mark for compilation tests to have an end-to-end test that would have "failed" without your change (we'd have made Stage depend on Unstage) but passes now?

trentlo replied:
Good suggestion. Added.

sanjoy reviewed (changes requested):
Looks really good Trent!

I only have some minor comments.

trentlo committed:
Rename _XlaAutoJitScope to _XlaInternalScope.

- Make it clear that the attribute is generated/used internally by auto_jit
and distinguish it from the user-provided _XlaScope.
- Also revise some other naming.

trentlo committed:
Add a new unittest in mark_for_compilation_pass_test.

The new test tests that ClusterScopingPass works and MarkForCompilationPass
accordinly preserves the required clustering scopes.

trentlo reviewed (commented):
@sanjoy, please see my replies and I should have addressed all of your comments. Please help to take a look one more time. Thanks a lot.

sanjoy commented on tensorflow/compiler/jit/cluster_scoping_pass.cc:
```diff
@@ -0,0 +1,148 @@
+  std::vector<Node*> starts;
+  starts.push_back(start);
+  auto enter = [&](Node* n) { AddOrAppendXlaInternalScope(n, unique_suffix); };
+  ReverseDFSFrom(*graph_, starts, enter, /*leave=*/nullptr,
```
Why do we not need an edge filter here?  We can have backwards cycles too right?

trentlo replied:
Yes, there can be backward cycles. The only reason there is not edge filter is because `ReverseDFSFromHelper` does not accept an edge filter as input argument (for reasons unclear to me). I should have brought up this for discussions but I somehow completely forgot this asymmetry. (Excuse me for this.)

On a second thought on the traversal, I am now towards traversing along backedges (i.e., do not exclude backedge). In the current _stage_ example, it is hard to imagine that we have both `stage` and `unstage` in the same loop. If they are not in the same loop, we should optimize towards having larger clusters and we should not break the loop into multiple clusters.

Hence, traversing along backedges should make more sense for the current case and we can re-visit this in the future when we see more cases. If this makes sense to you, I will update the code to remove edge filters (with some notes). Please feel free to let me know what you think?

trentlo replied:
Updated the codes as I convinced myself that DFS without filtering backedges is a more desirable behavior. Still, let me know what you think. Thanks.

sanjoy replied:
No filter SGTM.

sanjoy reviewed (changes requested):
Minor question inline.

trentlo committed:
[TF:XLA] Do not filter backedge in DFS traversal in cluster_scoping.

Do not filter any edges when traversing to find out all of the transitive
successor closure.  This should better serve the current needs to preserve the
pipeline stages by separating the transitive predecessors of Stage and
transitive successors of Unstage whenever possible.

sanjoy commented on tensorflow/compiler/jit/cluster_scoping_pass.cc:
```diff
@@ -0,0 +1,148 @@
+  if (cur_scope == absl::nullopt) {
+    updated_scope = std::string(suffix);
+  } else {
+    updated_scope = absl::StrCat(cur_scope.value(), "&", suffix);
```
Pedantically this is bit odd.  If we have `Unstage` -> `X` -> `Stage`; `Y` -> `Stage` then we'd be fine putting `X` and `Y` in the same cluster.  However, we won't do that in this scheme.

I think that's fine though (the situation above should be rare), but it would be nice to add a comment, if my understanding above is correct.

trentlo replied:
See the above comment for the response.

sanjoy reviewed (changes requested):
Another minor nit, but otherwise LGTM.

trentlo committed:
[TF:XLA] Polish the comments on AddOrAppendXlaInternalScope().

Polish the comments to better explain how the scope conjunction works in
the cluster scoping pass.

trentlo committed:
[TF:XLA] Minor formatting.

trentlo commented on tensorflow/compiler/jit/cluster_scoping_pass.cc:
```diff
@@ -0,0 +1,148 @@
+
+// NB! We append a new scope as suffix to the _XlaInternalScope attribute
+// instead of overriding the old value.  In this way, we respect the original
+// scopes.  In other words, appending X to Y creates the conjunction of the
```
I see where you are from. It is indeed confusing what it means to respect the "original" scope. And yes, getting both scopes is a good description.  I rewrote the comments by adopting your view and also giving a better explanation to my view. My view is that a unique scope in effect disallows merging a node to nodes in other scopes. So, scope conjunction preserves this semantic of disallowing merging.

In the below example, I'd like to point out that putting X and Y into different clusters is usually a more desirable result. Note that the long latency op is usually the `IteratorGetNext`, which processes input image with a lambda function. So, we do not want X to depend on IteratorGetNext.

`Unstage` -> `X` -> `Stage`;  `IteratorGetNext` -> `Y` -> `Stage`

On the other hand, `Stage`/`Unstage` themselves do not take a time too long to enqueue/dequeue. If you confuse them with long latency ops, it may be my fault as I (mistakenly) indicated so in a previous email. In contrast, Stage/Unstage is better treated as having semantics of wanted parallelism (instead of long latency.)

trentlo reviewed (commented):
I updated the comments and incorporated your example with explanations in the code.

Please help to take a look to see if this clarifies the issues. Thank you for the thorough review.

sanjoy reviewed (approved):
lgtm

mellanox-github commented:
Can one of the admins verify this patch?
